crates/ty_server/tests/e2e/pull_diagnostics.rs

```diff
@@ -27,3 +30,150 @@ def foo() -> str:
 
     Ok(())
 }
+
+#[test]
+fn workspace_diagnostic_caching() -> Result<()> {
+    // Redact result_id values since they are hash-based and non-deterministic
+    let mut settings = insta::Settings::clone_current();
+    settings.add_filter(r#""[a-f0-9]{16}""#, r#""[RESULT_ID]""#);
+    let _scoped = settings.bind_to_scope();
+
+    let workspace_root = SystemPath::new("src");
+
+    // File A: Will have an unchanged diagnostic
+    let file_a = SystemPath::new("src/unchanged.py");
+    let file_a_content = "\
+def foo() -> str:
+    return 42  # This error will remain the same
+";
+
+    // File B: Initially no error, will get a new error (added diagnostic)
+    let file_b = SystemPath::new("src/new_error.py");
+    let file_b_content_v1 = "\
+def foo() -> int:
+    return 42  # No error initially
+";
+    let file_b_content_v2 = "\
+def foo() -> str:
+    return 42  # Error appears
+";
+
+    // File C: Initially has error, will be fixed (removed diagnostic)
+    let file_c = SystemPath::new("src/fixed_error.py");
+    let file_c_content_v1 = "\
+def foo() -> str:
+    return 42  # Error initially
+";
+    let file_c_content_v2 = "\
+def foo() -> str:
+    return \"fixed\"  # Error removed
+";
+
+    // File D: Has error that changes content (changed diagnostic)
+    let file_d = SystemPath::new("src/changed_error.py");
+    let file_d_content_v1 = "\
+def foo() -> str:
+    return 42  # First error: expected str, got int
+";
+    let file_d_content_v2 = "\
+def foo() -> int:
+    return \"hello\"  # Different error: expected int, got str
+";
+
+    let global_options = ClientOptions::default().with_diagnostic_mode(DiagnosticMode::Workspace);
+
+    let mut server = TestServerBuilder::new()?
+        .with_workspace(
+            workspace_root,
+            ClientOptions::default().with_diagnostic_mode(DiagnosticMode::Workspace),
+        )?
```

Comment:
I wonder if this should actually be fixed in the test server instead to mimic what a client would do. This is because, I think, it's VS Code that does this part or more specifically the ty VS Code extension that creates the workspace and global settings and sends it via the initialization options.

The extension would construct workspace settings by asking VS Code the values for each setting which would automatically fallback to the global value. This means the server doesn't need to do any resolution between workspace and global values. This is what Ruff (and ty) does currently.

https://github.com/astral-sh/ty-vscode/blob/641781c780fad4d575c3ee715e2a6f70e5e29a8c/src/common/settings.ts

Reply:
Ideally, we'd fix this in the server itself, but it feels out of scope for this PR. I'm not entirely sure how I would model this but the server should prefer the option specified in the global option over a value provided in the workspace configuration. In fact, it should only use the workspace level option if it wasn't provided globally. But I'd prefer not to make changes to settings while you're also doing a refactor. That's why I'd prefer to leave this as is for now

Reply:
That sounds reasonable.